src/main/java/io/cdap/plugin/snowflake/source/batch/SnowflakeBatchSource.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.snowflake.source.batch;
+
+import io.cdap.cdap.api.annotation.Description;
+import io.cdap.cdap.api.annotation.Name;
+import io.cdap.cdap.api.annotation.Plugin;
+import io.cdap.cdap.api.data.batch.Input;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.api.dataset.lib.KeyValue;
+import io.cdap.cdap.etl.api.Emitter;
+import io.cdap.cdap.etl.api.FailureCollector;
+import io.cdap.cdap.etl.api.PipelineConfigurer;
+import io.cdap.cdap.etl.api.batch.BatchRuntimeContext;
+import io.cdap.cdap.etl.api.batch.BatchSource;
+import io.cdap.cdap.etl.api.batch.BatchSourceContext;
+import io.cdap.plugin.common.LineageRecorder;
+import io.cdap.plugin.snowflake.common.util.SchemaHelper;
+import org.apache.hadoop.io.NullWritable;
+
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Source plugin to read data from Snowflake.
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name(SnowflakeBatchSource.NAME)
+@Description("Read data from Snowflake.")
+public class SnowflakeBatchSource extends BatchSource<NullWritable, Map<String, String>, StructuredRecord> {
+
+  public static final String NAME = "Snowflake";
+
+  private final SnowflakeBatchSourceConfig config;
+  private SnowflakeMapToRecordTransformer transformer;
+
+  public SnowflakeBatchSource(SnowflakeBatchSourceConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    FailureCollector failureCollector = pipelineConfigurer.getStageConfigurer().getFailureCollector();
+    config.validate(failureCollector);
+
+    SnowflakeSourceAccessor snowflakeAccessor = new SnowflakeSourceAccessor(config);
+    Schema schema = SchemaHelper.getSchema(snowflakeAccessor, config.getSchema(),
+                                           failureCollector, config.getImportQuery());
```

Comment:
If both schema and import query are macros, you can't get the output schema in configurePipeline.
Please test with macros in all fields.

Reply:
Tested it and updated the patch. @rmstar Can you please review/approve? Thanks.